### PR TITLE
STM - States and Transitions

### PR DESCRIPTION
**Pull request by DH254**
Updated states to match the 2024 plans for the state machine. 
Updated tests to match.

### CODE REVIEW

File: lib/state_machine/message.hpp
Comment (TomLonergan03) [inline lib/state_machine/message.hpp:1]: Is there a reason to duplicate this stuff across 2 files? I assume this was for ROS originally

File: lib/state_machine/state.hpp [hunk 1]
```diff
@@ -5,18 +5,20 @@ namespace hyped::state_machine {
 enum class State {
   kIdle,
   kCalibrating,
```
Comment (TomLonergan03) [inline lib/state_machine/state.hpp:7]: nitpick but stick to one case, either `calibrating`, `accelerating` etc or `calibrate`, `accelerate`

File: lib/state_machine/state_machine.cpp [hunk 2]
```diff
@@ -23,4 +23,14 @@ State StateMachine::getCurrentState()
   return current_state_;
 }
 
+core::Result StateMachine::handleMessage(const Message &message)
+{
+  const auto temp = transition_to_state_.find({current_state_, message});
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.cpp:28]: `next_state` maybe? at least something more descriptive than `temp`

(nice work on the `const` btw)

File: lib/state_machine/state.hpp [hunk 3]
```diff
@@ -5,18 +5,20 @@ namespace hyped::state_machine {
...
-  kAccelerating,
-  kCruising,
-  kMotorBraking,
-  kPreFrictionBraking,
-  kFrictionBraking,
-  kPreFrictionBrakingFail,
-  kFrictionBrakingFail,
-  kFailureBraking,
+  kAccelerate,
+  kLIMBrake,
```
Comment (TomLonergan03) [inline lib/state_machine/state.hpp:13]: `kLimBrake`

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]
Overall very good, especially for a first PR! Mostly just nitpicks

File: lib/state_machine/state_machine.hpp [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #include <queue>
 #include <unordered_map>
 
+#include "core/types.hpp"
```
Comment (DylanCavers) [inline lib/state_machine/state_machine.hpp:10]: Pedantic but can you move this to where the other hyped imports are

File: test/state_machine/stm_tests.cpp [hunk 5]
```diff
@@ -4,9 +4,170 @@
...
+  testTransition(
+    stm, state_machine::Message::kCapacitorDischarge, state_machine::State::kCapacitorDischarge);
+  testTransition(stm, state_machine::Message::kSafe, state_machine::State::kSafe);
+  testTransition(stm, state_machine::Message::kSafe, state_machine::State::kSafe);
 }
+}  // namespace hyped::test
 
-}  // namespace hyped::test
\ No newline at end of file
+// namespace hyped::test
```
Comment (DylanCavers) [inline test/state_machine/stm_tests.cpp:173]: and remove duplicated namespace

File: test/state_machine/stm_tests.cpp [hunk 6]
```diff
@@ -5,169 +5,164 @@
 namespace hyped::test {
 
 void testTransition(std::unique_ptr<state_machine::StateMachine> &stm,
```
Comment (TomLonergan03) [inline test/state_machine/stm_tests.cpp:7]: why is this a reference to a `unique_ptr`?

File: test/state_machine/stm_tests.cpp
Comment (TomLonergan03) [inline test/state_machine/stm_tests.cpp:1]: test for duplicate messages occurring after a transition?

e.g. calibrate -> precharge -> calibrate -> ready for lev

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: test/state_machine/stm_tests.cpp (hunk 6 above)
Comment (DylanCavers) [reply test/state_machine/stm_tests.cpp:7]: I think this is from what I wrote last year, can just be changed

File: test/state_machine/stm_tests.cpp [hunk 7]
```diff
@@ -4,9 +4,175 @@
...
+                    state_machine::State expected_state)
 {
-  std::unique_ptr stm = std::make_unique<state_machine::StateMachine>();
+  stm->handleTransition(transition_state);
+  ASSERT_TRUE(stm->getCurrentState() == expected_state);
 }
 
+TEST(StateMachine, cleanRun)
+{
+  std::shared_ptr stm = std::make_shared<state_machine::StateMachine>();
```
Comment (DylanCavers) [inline test/state_machine/stm_tests.cpp:17]: probably use `auto` here

**Review by DylanCavers — APPROVED** [review]
looks good to me

File: lib/state_machine/state_machine.hpp [hunk 8]
```diff
@@ -8,68 +8,75 @@
...
-       {Message::kPreFrictionBrake, "kPreFrictionBrake"},
-       {Message::kFrictionBrake, "kFrictionBrake"},
-       {Message::kStopped, "kStopped"},
-       {Message::kFailureBrake, "kFailureBrake"},
-       {Message::kPreFrictionBrakeFail, "kPreFrictionBrakeFail"},
-       {Message::kFrictionBrakeFail, "kFrictionBrakeFail"},
-       {Message::kOff, "kOff"},
-       {Message::kFailureStopped, "kFailureStopped"},
-       {Message::kFailureOff, "kFailureOff"}};
+  const std::unordered_map<std::string, State> string_to_state_
```
Comment (TomLonergan03) [inline lib/state_machine/state_machine.hpp:24]: is there a reason all these hashmaps aren't `static`?

File: lib/state_machine/types.hpp [hunk 9]
```diff
@@ -10,7 +9,7 @@
 namespace hyped::state_machine {
 struct SourceAndMessage {
   State source;
-  Message message;
+  State message;
```
Comment (TomLonergan03) [inline lib/state_machine/types.hpp:12]: `target` maybe?

**Review by TomLonergan03 — CHANGES_REQUESTED** [review]

File: lib/state_machine/state_machine.hpp (hunk 8 above)
Comment (TomLonergan03) [reply lib/state_machine/state_machine.hpp:24]: (this isn't a strong feeling, I'm happy to hear arguments otherwise)

**Review by TomLonergan03 — APPROVED** [review]
Looks good!

**Pull request merged**